overlays/nodejs/s3/s3Mixins.ts

```diff
@@ -0,0 +1,232 @@
+    event?: "*" | "Delete" | "DeleteMarkerCreated";
+}
+
+// See https://docs.aws.amazon.com/AmazonS3/latest/dev/notification-content-structure.html.
+export interface BucketEvent {
+    Records?: BucketRecord[];
+}
+
+export interface BucketRecord {
+    eventVersion: string;
+    eventSource: string;
+    awsRegion: string;
+    eventTime: string;
+    eventName: string;
+    userIdentity: {
+        principalId: string;
+    };
+    requestParameters: {
+        sourceIPAddress: string;
+    };
+    responseElements: {
+        "x-amz-request-id": string;
+        "x-amz-id-2": string;
+    };
+    s3: {
+        s3SchemaVersion: string;
+        configurationId: string;
+        bucket: {
+            name: string;
+            ownerIdentity: {
+                principalId: string;
+            },
+            arn: string;
+        };
+        object: {
+            key: string;
+            size: number;
+            eTag: string;
+            versionId?: string;
+            sequencer: string;
+        };
+    };
+}
+
+export type BucketEventHandler = lambda.EventHandler<BucketEvent, void>;
+
+interface SubscriptionInfo {
+    name: string;
+    events: string[];
+    filterPrefix?: string;
+    filterSuffix?: string;
+    lambdaFunctionArn: pulumi.Output<string>;
+    permission: lambda.Permission;
+}
+
+let bucketSubscriptionInfos = new Map<Bucket, SubscriptionInfo[]>();
+
+/**
+ * A component corresponding to a single underlying aws.s3.BucketNotification created for a bucket.
+ * Note: due to the AWS requirement that all notifications for a bucket be defined at once, the
+ * actual aws.s3.BucketNotification instances will only be created once the pulumi program runs to
+ * completion and all subscriptions have been heard about.
+ */
+export class BucketEventSubscription extends lambda.EventSubscription {
+    public readonly bucket: pulumi.Output<Bucket>;
+
+    public constructor(
+        name: string, bucket: Bucket, handler: BucketEventHandler,
+        args: BucketEventSubscriptionArgs, opts?: pulumi.ResourceOptions) {
+
+        super("aws:s3:BucketEventSubscription", name, { bucket: bucket }, opts);
+
+        this.func = lambda.createFunctionFromEventHandler(name, handler, { parent: this });
+
+        const permission = new lambda.Permission(name, {
+            function: this.func,
+            action: "lambda:InvokeFunction",
+            principal: "s3.amazonaws.com",
+            sourceArn: bucket.id.apply(bucketName => `arn:aws:s3:::${bucketName}`),
+        }, { parent: this });
+
+        this.permission = permission;
+
+        // We must create only a single BucketNotification per Bucket per AWS API limitations.  See
+        // https://github.com/terraform-providers/terraform-provider-aws/issues/1715.  So we push
+        // the subscription information here, and then actually create the BucketNotification if
+        // needed on process `beforeExit`.
```

Comment:
Eek, I guess this is ok -- but kind of scary to create resources just before the process exits like this.

Reply:
We've been doing this for months.  This is how Bucket.put has had to work since day 1.  It actually hasn't been a problem for us at all :)